lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,143 @@
+      const messages = errorMessageBuilder.buildErrorMessages(validate.errors, userConfig);
+      this.handleErrorMessages(messages);
+    }
+  }
+
+  handleErrorMessages(messages) {
+    for (const message of messages) {
+      switch (this.serverless.service.configValidationMode) {
+        case 'error':
+          throw new this.serverless.classes.Error(`${ERROR_PREFIX} ${messages.join('; ')}`);
+        case 'warn':
+          this.serverless.cli.log(`${ERROR_PREFIX} ${message}`, 'Serverless', { color: 'red' });
+          break;
+        default:
+          throw new this.serverless.classes.Error(
+            `${ERROR_PREFIX} .configValidationMode should be equal to one of the allowed values 'error' or 'warn'`
+          );
+      }
+    }
+  }
+
+  defineTopLevelProperty(name, subSchema) {
+    this.schema.properties[name] = subSchema;
+  }
+
+  defineProvider(name, options = {}) {
+    // TODO: Remove this 'if' statement and place this piece of schema inside
+    // initial schema. This condition was added because some other tests
+    // were failing, see https://travis-ci.org/github/serverless/serverless/jobs/663153831
+    if (!this.schema.properties.provider.properties.name) {
+      this.schema.properties.provider.properties.name = { enum: [] };
+    }
+
+    this.schema.properties.provider.properties.name.enum.push(name);
+
+    if (options.provider) {
+      addPropertiesToSchema(this.schema.properties.provider, options.provider);
+    }
+
+    if (options.function) {
+      addPropertiesToSchema(
+        this.schema.properties.functions.patternProperties[FUNCTION_NAME_PATTERN],
+        options.function
+      );
+    }
+  }
+
+  defineCustomProperties(configSchemaParts) {
+    addPropertiesToSchema(this.schema.properties.custom, configSchemaParts);
+  }
+
+  defineFunctionEvent(name, configSchema) {
+    if (
+      !this.schema.properties.functions.patternProperties[FUNCTION_NAME_PATTERN].properties.events
+        .items.anyOf
+    ) {
+      this.schema.properties.functions.patternProperties[
+        FUNCTION_NAME_PATTERN
+      ].properties.events.items.anyOf = [];
+    }
+
+    this.schema.properties.functions.patternProperties[
+      FUNCTION_NAME_PATTERN
+    ].properties.events.items.anyOf.push({
+      type: 'object',
+      properties: { [name]: configSchema },
+      required: [name],
+      additionalProperties: false,
+    });
+  }
+}
+
+function addPropertiesToSchema(subSchema, extension = { properties: {}, required: [] }) {
+  subSchema.properties = Object.assign(subSchema.properties, extension.properties);
+
+  if (!subSchema.required) subSchema.required = [];
+
+  if (Array.isArray(extension.required)) subSchema.required.push(...extension.required);
+}
+
+function deepFreeze(object) {
+  // Retrieve the property names defined on object
+  const propNames = Object.getOwnPropertyNames(object);
+
+  // Freeze properties before freezing self
```

Comment:
I guess order is irrelevant, isn't it?

Reply:
Yes, it is. When copy-pasting code from the Internet I tend to leave it as is. But you are right, I'll remove this comment